File: app/src/androidTest/java/com/example/all_sensors/tests/HealthTrackingTestClient.java
package com.example.all_sensors.tests;

import android.content.Context;
import android.util.Log;

import com.samsung.android.service.health.tracking.ConnectionListener;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.HealthTrackerCapability;
import com.samsung.android.service.health.tracking.HealthTrackerException;
import com.samsung.android.service.health.tracking.HealthTrackingService;
import com.samsung.android.service.health.tracking.data.HealthTrackerType;
import com.samsung.android.service.health.tracking.data.PpgType;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Cliente ligero para las pruebas instrumentadas que se encarga de conectar el SDK,
 * exponer la capability y crear trackers con la configuración necesaria.
 */
class HealthTrackingTestClient {
    private static final String TAG = "HealthTrackingTestClient";
    private final HealthTrackingService service;
    private final CountDownLatch connectedLatch = new CountDownLatch(1);
    private final List<HealthTracker> startedTrackers = new ArrayList<>();
    private volatile boolean connected = false;

    HealthTrackingTestClient(Context context) {
        service = new HealthTrackingService(new ConnectionListener() {
            @Override
            public void onConnectionSuccess() {
                connected = true;
                connectedLatch.countDown();
            }

            @Override
            public void onConnectionFailed(HealthTrackerException e) {
                connectedLatch.countDown();
                Log.e(TAG, "Fallo al conectar con HealthTrackingService", e);
            }

            @Override
            public void onConnectionEnded() {
                connected = false;
            }
        }, context);
    }

    void connectOrThrow(long timeoutMs) throws InterruptedException {
        service.connectService();
        if (!connectedLatch.await(timeoutMs, TimeUnit.MILLISECONDS) || !connected) {
            throw new IllegalStateException("No se pudo conectar con HealthTrackingService");
        }
    }

    HealthTrackerCapability getCapability() {
        ensureConnected();
        return service.getTrackingCapability();
    }

    HealthTracker startTracking(HealthTrackerType type, HealthTracker.TrackerEventListener listener) {
        ensureConnected();
        HealthTracker tracker = buildTracker(type);
        tracker.setEventListener(listener);
        startedTrackers.add(tracker);
        return tracker;
    }

    void stopAll() {
        for (HealthTracker tracker : startedTrackers) {
            tracker.unsetEventListener();
        }
        startedTrackers.clear();
        if (connected) {
            service.disconnectService();
            connected = false;
        }
    }

    private HealthTracker buildTracker(HealthTrackerType type) {
        // Para los PPG permitimos pasar una lista explícita de tipos de LED.
        Set<PpgType> ppgTypes = EnumSet.of(PpgType.GREEN);
        switch (type) {
            case PPG_CONTINUOUS:
            case PPG_ON_DEMAND:
            case PPG_GREEN:
            case PPG_RED:
            case PPG_IR:
                return service.getHealthTracker(type, ppgTypes);
            default:
                return service.getHealthTracker(type);
        }
    }

    private void ensureConnected() {
        if (!connected) {
            throw new IllegalStateException("El servicio no está conectado");
        }
    }
}

File: app/src/androidTest/java/com/signals/smartwatch/tests/HealthTrackingTestClient.kt
package com.signals.smartwatch.tests

import android.content.Context
import android.util.Log
import com.samsung.android.service.health.tracking.ConnectionListener
import com.samsung.android.service.health.tracking.HealthTracker
import com.samsung.android.service.health.tracking.HealthTrackerCapability
import com.samsung.android.service.health.tracking.HealthTrackerException
import com.samsung.android.service.health.tracking.HealthTrackingService
import com.samsung.android.service.health.tracking.data.HealthTrackerType
import com.samsung.android.service.health.tracking.data.PpgType
import java.util.EnumSet
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Cliente ligero para pruebas instrumentadas: conecta el SDK, expone la capability
 * y crea trackers con la configuración necesaria.
 */
internal class HealthTrackingTestClient(context: Context) {

    private val connectedLatch = CountDownLatch(1)
    private val startedTrackers = mutableListOf<HealthTracker>()

    @Volatile private var connected = false

    private val service = HealthTrackingService(object : ConnectionListener {
        override fun onConnectionSuccess() {
            connected = true
            connectedLatch.countDown()
        }
        override fun onConnectionFailed(e: HealthTrackerException) {
            Log.e(TAG, "Fallo al conectar con HealthTrackingService", e)
            connectedLatch.countDown()
        }
        override fun onConnectionEnded() { connected = false }
    }, context)

    fun connectOrThrow(timeoutMs: Long) {
        service.connectService()
        check(connectedLatch.await(timeoutMs, TimeUnit.MILLISECONDS) && connected) {
            "No se pudo conectar con HealthTrackingService"
        }
    }

    fun getCapability(): HealthTrackerCapability {
        ensureConnected()
        return service.trackingCapability
    }

    fun startTracking(type: HealthTrackerType, listener: HealthTracker.TrackerEventListener): HealthTracker {
        ensureConnected()
        return buildTracker(type).also {
            it.setEventListener(listener)
            startedTrackers += it
        }
    }

    fun stopAll() {
        startedTrackers.forEach { it.unsetEventListener() }
        startedTrackers.clear()
        if (connected) {
            service.disconnectService()
            connected = false
        }
    }

    private fun buildTracker(type: HealthTrackerType): HealthTracker {
        val ppgTypes = EnumSet.of(PpgType.GREEN)
        return when (type) {
            HealthTrackerType.PPG_CONTINUOUS,
            HealthTrackerType.PPG_ON_DEMAND,
            HealthTrackerType.PPG_GREEN,
            HealthTrackerType.PPG_RED,
            HealthTrackerType.PPG_IR -> service.getHealthTracker(type, ppgTypes)
            else                     -> service.getHealthTracker(type)
        }
    }

    private fun ensureConnected() = check(connected) { "El servicio no está conectado" }

    companion object { private const val TAG = "HealthTrackingTestClient" }
}
